app/controllers/application_controller.rb

```diff
@@ -60,8 +61,9 @@ def authenticate_token
 
   def authenticate_keycard
     @current_user = User.new
-    @current_user.identity = Keycard::RequestAttributes.new(request)
-    if @current_user.identity[:username].empty?
+    @current_user.identity = Keycard::Request::AttributesFactory.new.for(request).identity
+    @current_user.identity[:username] = @current_user.identity[:user_pid]
```

Comment:
This won't do anything for token auth (the user_pid is not set). We should trace what the dependency on `:username` in the identity is -- recent Keycard does not use that key at all, so I think it's vestigial.